bin/getstatus.py

```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/python3
+
+##################################################
+#    This file is part of the HABTracker project for tracking high altitude balloons.
+#
+#    Copyright (C) 2019, Jeff Deaton (N6BA)
+#
+#    HABTracker is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    HABTracker is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with HABTracker.  If not, see <https://www.gnu.org/licenses/>.
+#
+##################################################
+
+import psutil
+import sys
+import os
+import json
+
+
+
+########################
+# getProcStatus
+# searches for the processes listed in 'query' among all procs running on the system.  It returns a list of of those found
+########################
+def getProcStatus(query: list)->list:
+
+    # where we store the processes and their status
+    procstatus = []
+
+    # Pre-populate the processes by looping through those in the 'query' list
+    for p in query:
+        procstatus.append({"process": p, "status": 0, "cmdline": None })
+
+    # Iterate over all running processes, searching for those in 'procstatus'
+    for proc in psutil.process_iter():
+
+       try:
+           # Get process detail as dictionary
+           pInfoDict = proc.as_dict(attrs=['pid', 'ppid', 'name', 'exe', 'memory_percent', 'cmdline' ])
+
+       except (psutil.NoSuchProcess, psutil.AccessDenied):
+           # we just skip over those processes that we're not allowed to see or if they no longer exist.
+           pass
+
+       else:
+           # for everyother process, we check if it's one of those that we're looking for
+           # loop over each process in procstatus checking if the current OS process is a match.  When a match is found we break out of the loop.
+           for p in procstatus:
+               if pInfoDict["name"] and pInfoDict["cmdline"]:
+                   if p["process"] in pInfoDict["name"].lower() or p["process"] in pInfoDict["cmdline"]:
+                       p["cmdline"] = pInfoDict["cmdline"]
+                       p["status"] = 1  
+                       break
```

Comment:
I looked at this whole block for several minutes wondering if there was a clever way to reduce the lookup time by storing the procstatus in an associative array or sorted list. I couldn't come up with anything. Either the setup cost would outweigh the benefit or the complexity of searching on two different fields made it too cumbersome. What you have is simple and direct.

Reply:
Yeah, agreed.  I’m sure it’s inefficient and I started to look at this for a while as this is called every 5 seconds from the web interface trying to determine what processes are running.  Having a better way to reduce the lookup time would be good.  This code has been around since the beginning and was likely borrowed from an example on the web somewhere.  It’d probably be a worthwhile endeavor to step back and look at the bigger picture of how to determine if a process is running or not.